File: typescript/src/ancilis/cli/validate.ts
/** ancilis config validate — config validation with actionable errors. */

import { inspectConfigFileMigration, loadConfig, migrateConfigFile } from "../config/index.js";

export interface ConfigValidateOptions {
  verbose?: boolean;
}

export interface ConfigMigrateOptions {
  apply?: boolean;
}

export function validateAndFormat(configPath?: string, options: ConfigValidateOptions = {}): { valid: boolean; message: string } {
  const lines: string[] = [];

  try {
    const migration = configPath ? inspectConfigFileMigration(configPath) : null;
    const raw = configPath ? { path: configPath } : undefined;
    const resolved = loadConfig(raw ?? {});

    lines.push("\u2713 Config valid");
    lines.push(`  Config version: ${resolved.configVersion ?? "unknown"}`);
    lines.push(`  Agent: ${resolved.agentName}`);
    lines.push(`  Mode: ${resolved.mode}`);

    if (migration?.changed) {
      lines.push(`  Migration available: v${migration.originalVersion} -> v${migration.currentVersion}`);
      for (const change of migration.changes) {
        lines.push(`    - ${change}`);
      }
      lines.push("    Run `ancilis config migrate --apply` to write the migrated config and backup.");
    }

    // Activation summary
    const activationLines: string[] = [];
    for (const certId of resolved.activeCertifications) {
      const enabledCount = [...resolved.controls.values()].filter(c => c.enabled).length;
      activationLines.push(`    certification_targets: [${certId}] \u2192 ${certId.toUpperCase()} active, ${enabledCount} controls`);
    }
    for (const [, oa] of [...resolved.activeOverlays.entries()].sort((a, b) => a[0].localeCompare(b[0]))) {
      const trigger = oa.triggeredBy.length > 0 ? oa.triggeredBy.join(", ") : "explicit";
      activationLines.push(`    ${oa.name} overlay active (triggered by ${trigger})`);
    }

    if (activationLines.length > 0) {
      lines.push("  Activation:");
      lines.push(...activationLines);
    }

    const enabled = [...resolved.controls.values()].filter(c => c.enabled).length;
    lines.push(`  Controls: ${enabled} active`);

    if (resolved.warnings.length > 0) {
      lines.push("  Warnings:");
      for (const warning of resolved.warnings) {
        lines.push(`    - ${warning}`);
      }
    }

    if (options.verbose) {
      lines.push("  Schema:");
      lines.push("    shared/schemas/config.schema.json");
      lines.push("  Active overlays:");
      if (resolved.activeOverlays.size === 0) {
        lines.push("    - none");
      } else {
        for (const [overlayId, overlay] of [...resolved.activeOverlays.entries()].sort((a, b) => a[0].localeCompare(b[0]))) {
          const trigger = overlay.triggeredBy.length > 0 ? overlay.triggeredBy.join(", ") : "manual";
          lines.push(`    - ${overlayId}: ${overlay.name} (${trigger})`);
        }
      }
      lines.push("  Enabled controls:");
      for (const control of enabledControls(resolved)) {
        lines.push(`    - ${control.controlId}: ${control.name}`);
      }
    }

    return { valid: true, message: lines.join("\n") };
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    lines.push("\u2717 Config invalid");
    lines.push(`  ${msg}`);
    return { valid: false, message: lines.join("\n") };
  }
}

function enabledControls(resolved: ReturnType<typeof loadConfig>): Array<{ controlId: string; name: string }> {
  return [...resolved.controls.values()]
    .filter(control => control.enabled)
    .sort((a, b) => a.controlId.localeCompare(b.controlId))
    .map(control => ({ controlId: control.controlId, name: control.name }));
}

export function migrateAndFormat(configPath: string, options: ConfigMigrateOptions = {}): { ok: boolean; message: string } {
  const lines: string[] = [];

  try {
    const result = migrateConfigFile(configPath, { apply: options.apply });
    if (!result.changed) {
      return {
        ok: true,
        message: `Config already at version ${result.currentVersion}; no migration needed.`,
      };
    }

    lines.push(
      options.apply
        ? `Migrated ${configPath} from v${result.originalVersion} to v${result.currentVersion}.`
        : `Preview migration for ${configPath}: v${result.originalVersion} -> v${result.currentVersion}.`,
    );
    for (const change of result.changes) {
      lines.push(`  - ${change}`);
    }
    if (options.apply) {
      lines.push(`Backup written to ${result.backupPath}`);
    } else {
      lines.push("Run `ancilis config migrate --apply` to write these changes.");
    }

    return { ok: true, message: lines.join("\n") };
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    return { ok: false, message: `Config migration failed\n  ${msg}` };
  }
}

File: python/src/ancilis/cli/validate.py
"""ancilis config validate — config validation with actionable errors."""

from __future__ import annotations

# mypy: disable-error-code=import-untyped

import click
import yaml

from ancilis.config import VALID_CONTROL_IDS, load_config, load_taxonomy
from ancilis.errors import ConfigError


VALID_CERT_TARGETS = ["aiuc-1"]


def _validate_and_format(config_path: str | None) -> tuple[bool, str]:
    """Validate config and return (valid, message)."""
    lines: list[str] = []

    try:
        resolved = load_config(path=config_path) if config_path else load_config()
    except FileNotFoundError:
        return False, "\u2717 Config not found\n  No ancilis.yaml found in current directory.\n  Create one with: agent.name set to your agent's name."
    except (ValueError, ConfigError) as e:
        msg = str(e)
        lines.append("\u2717 Config invalid")
        lines.append(f"  {msg}")

        # Add actionable hints
        if "Unknown data type" in msg:
            taxonomy = load_taxonomy()
            valid_types = sorted(taxonomy["developer_type_mapping"].keys())
            lines.append(f"  Available types: {', '.join(valid_types)}")
        elif "Unknown control ID" in msg:
            lines.append(f"  Available controls: {', '.join(sorted(VALID_CONTROL_IDS))}")
        elif "agent.name" in msg:
            lines.append("  Fix: add 'agent: { name: my-agent }' to ancilis.yaml")

        return False, "\n".join(lines)
    except Exception as e:
        return False, f"\u2717 Config invalid\n  {e}"

    lines.append("\u2713 Config valid")
    lines.append(f"  Agent: {resolved.agent_name}")
    lines.append(f"  Mode: {resolved.mode}")

    # Activation summary
    activation_lines: list[str] = []
    if resolved.active_certifications:
        for cert_id in resolved.active_certifications:
            enabled_count = sum(1 for c in resolved.controls.values() if c.enabled)
            activation_lines.append(
                f"    certification_targets: [{cert_id}] \u2192 {cert_id.upper()} active, {enabled_count} controls"
            )
    if resolved.active_overlays:
        for _oid, oa in sorted(resolved.active_overlays.items()):
            trigger = ", ".join(oa.triggered_by) if oa.triggered_by else "explicit"
            activation_lines.append(f"    {oa.name} overlay active (triggered by {trigger})")

    if activation_lines:
        lines.append("  Activation:")
        lines.extend(activation_lines)

    enabled = sum(1 for c in resolved.controls.values() if c.enabled)
    strict = [
        c for c in resolved.controls.values()
        if c.enabled and c.threshold == "strict"
    ]
    controls_detail = f"  Controls: {enabled} active"
    if strict:
        strict_names = ", ".join(
            f"{c.control_id} at strict threshold"
            for c in strict
        )
        controls_detail += f" ({strict_names})"
    lines.append(controls_detail)

    # Unavailable overlays (roadmap)
    if resolved.unavailable_overlays:
        lines.append("  Roadmap overlays:")
        for uo in resolved.unavailable_overlays:
            lines.append(
                f"    ? {uo.overlay_id} would be activated by {uo.triggered_by} "
                f"via {uo.data_type} but is not yet available"
            )
        lines.append("    Baseline security controls are active for all data types.")

    if resolved.warnings:
        lines.append("  Warnings:")
        for w in resolved.warnings:
            lines.append(f"    ! {w}")

    return True, "\n".join(lines)


@click.command()
@click.argument("config_path_arg", required=False)
@click.option("--config", "config_path_option", default=None, help="Path to ancilis.yaml")
def validate(config_path_arg: str | None, config_path_option: str | None) -> None:
    """Validate ancilis.yaml configuration."""
    config_path = config_path_option if config_path_option is not None else config_path_arg
    valid, message = _validate_and_format(config_path)
    click.echo(message)
    if not valid:
        raise SystemExit(1)
